node/http.go

```diff
@@ -112,6 +113,70 @@ func makeRPCOverWebsocket(host string, port uint16, servers map[string]*jsonrpc.
 	return makeHTTPService(host, port, cors.Default().Handler(mux))
 }
 
+type ipcService struct {
+	handlers []*jsonrpc.Ipc
+}
+
+var _ service.Service = (*ipcService)(nil)
+
+// Run is a method of the ipcService struct that starts the execution of IPC handlers in parallel.
+// It takes context as a parameter to control the execution and returns an error if any handler fails.
+func (ipc *ipcService) Run(ctx context.Context) error {
+	errCh := make(chan error, len(ipc.handlers))
+	defer close(errCh)
+
+	var wg conc.WaitGroup
+	defer wg.Wait()
+
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	// Iterate over all registered handlers and execute them concurrently.
+	for i := range ipc.handlers {
+		handler := ipc.handlers[i]
+		wg.Go(func() {
+			cancel(handler.Run(ctx))
+		})
+	}
+
+	<-ctx.Done()
+	return context.Cause(ctx)
+}
+
+func makeRPCOverIpc(dir string, servers map[string]*jsonrpc.Server,
+	log utils.SimpleLogger, metricsEnabled bool,
+) (*ipcService, error) {
+	var listener jsonrpc.NewRequestListener
+	if metricsEnabled {
+		listener = makeIpcMetrics()
+	}
+
+	srv := &ipcService{handlers: make([]*jsonrpc.Ipc, 0)}
+	for path := range servers {
+		server := servers[path]
+		trimmed, ok := strings.CutPrefix(path, "/")
+		if !ok {
+			continue
+		}
+		endpoint := filepath.Join(dir, fmt.Sprintf("juno_%s.ipc", trimmed))
+		if trimmed == "" {
+			endpoint = filepath.Join(dir, "juno.ipc")
+		}
+		ipcHandler, err := jsonrpc.NewIpc(server, endpoint, log)
+		if err != nil {
+			// ugly hack to not leave stale sockets.
+			ctx, cancel := context.WithCancel(context.Background())
+			cancel()
+			ipcHandler.Run(ctx) //nolint:errcheck
+			return nil, err
```

Comment:
Is this necessary?

Reply:
Yes, there could be other sockets spawned before that error and they should be cleaned up.

Reply:
Since we clean up stale sockets on startup, do you think it's ok to delete this code?

Reply:
The cleanup at startup is racy racy  since it's the same pattern as used [here](https://github.com/tailscale/tailscale/blob/e78cb9aeb3b32ee97991710c72eb979ebad61477/safesocket/unixsocket.go#L27-L45)

It's always better to ensure that resources are freed by the same process that's taking them.

Reply:
Does that mean we should delete the endpoints for the other `srv.handlers` as well?

Reply:
Yes and that's what's being done by those few lines.